### DISCUSSION TITLE
Non string object keys (CBOR)

### DISCUSSION THREAD

**Discussion by Pique7** [danielaparker/jsoncons#568] [Q&A]
Are numeric keys supported/possible (especially in CBOR serialization)?
Example:
```
{
  1: "value1",
  2: "value2"
}
```
I haven't found a solution yet.

**Comment by danielaparker** [comment]
For writing CBOR, if using [cbor_bytes_encoder or cbor_stream_encoder](https://github.com/danielaparker/jsoncons/blob/master/doc/ref/cbor/basic_cbor_encoder.md), you can write a key as any CBOR value, numeric, object, whatever, basically write pairs of two values (the first the key, the second the corresponding value) bracketed inside a `begin_object` and an `end_object`. Similarly with MessagePack.

For reading CBOR, we have `cbor_event_reader`, you can get an idea how it works from the [test cases](https://github.com/danielaparker/jsoncons/blob/master/test/cbor/src/cbor_event_reader_tests.cpp), basically it allows you to receive a stream of CBOR events including value/value pairs for objects (CBOR maps). These tests cases show string keys, but it can handle other types  too.

When interacting with the rest of the jsoncons library, the keys currently get converted to/from strings, though. For example, you can use `decode_cbor` to read CBOR bytes into a `std::map` with an `uint64_t` key,  but the key would first be converted into a string and then back to a `uint64_t`. This follows from the fact that the jsoncons pipeline that is common to all our data formats assumes string keys. This is something I'd like to revisit sometime.

**Reply by Pique7** [reply]
> ..., you can write a key as any CBOR value, numeric, object, whatever, basically write pairs of two values (the first the key, the second the corresponding value) bracketed inside a `begin_object` and an `end_object`.

OK, if I understand correctly, you mean e.g.:
```c
encoder.begin_object();
encoder.uint64_value(1);
encoder.string_value("value1");
encoder.uint64_value(2);
encoder.string_value("value2");
encoder.end_object();
```
Indeed, this works. 
But how can I specify a predefined length for the object? For example:
```c
encoder.begin_object(2);
...
encoder.end_object();
```

**Reply by danielaparker** [reply]
An object with precisely two key/value pairs:

```cpp
encoder.begin_object(2);
encoder.uint64_value(1);
encoder.string_value("value1");
encoder.uint64_value(2);
encoder.string_value("value2");
encoder.end_object();
```

**Reply by Pique7** [reply]
Hmm, I get an error `cbor_errc::too_many_items` on `encoder.end_object()`. Have I missed something?
The program executes without errors with `begin_object(4)` instead of `begin_object(2)` but in this case the resulting data is not valid CBOR.

**Reply by danielaparker** [reply]
@Pique7, you're right, that's a defect, is the case for msgpack too.  Fixed on master.

**Reply by Pique7** [reply]
Thanks for the speedy reaction and solution, but I use vcpkg and it still installs version 0.175.0 by default. I am able to update to v0.177.0 by manually overriding, but not to a later release.

**Reply by danielaparker** [reply]
1.0.0 should be available on vcpkg soon, just submitted a PR, should be available in a few days. It will include these fixes.
